### ISSUE TITLE
Regression with setting focus in test environment

### ISSUE THREAD

**Issue by nwhittaker** [Esri/calcite-design-system#7740]
### Check existing issues

- [X] I have [checked for existing issues](https://github.com/Esri/calcite-design-system/issues) to avoid duplicates

### Actual Behavior

In an automated test, if multiple components are given focus in quick succession, there's a race condition as to which one will actually retain focus.

### Expected Behavior

In an automated test, if multiple components are given focus in quick succession, the last component to gain focus retains it.

### Reproduction Sample

https://codepen.io/nwhittaker-esri/pen/eYbRGqP

### Reproduction Steps

1. Visit code sample
2. Click the **Set focus** button and see the first input retains focus despite the 2nd input being focused after.

### Reproduction Version

1.8.0

### Relevant Info

The code sample aims to mimic how an automated test interacts with components. Specifically via dispatched events instead of calling and awaiting `setFocus()` directly.

The issue appears to stem from changes in https://github.com/Esri/calcite-design-system/pull/7255 and https://github.com/Esri/calcite-design-system/pull/7277.

### Regression?

1.4.3

### Priority impact

p2 - want for current milestone

### Impact

This issue represents a significant amount of effort for the Field Maps team to debug and rewrite tests in order to upgrade beyond Calcite 1.4.3.

### Calcite package

- [X] @esri/calcite-components
- [ ] @esri/calcite-components-react

### Esri team

ArcGIS Field Apps

**Comment by jcfranco** [comment]
We'll have to do some exploratory work here first, but we might be able to help here by avoiding calls to `setFocus` internally when the host is clicked in some cases. As you probably saw in the issues you referenced, our component's rely  on focus being async to make sure the component is rendered before we can do any reliable internal focusing.

For your use case, is it mainly `calcite-input` wreaking havoc in tests or are there other components that are causing issues? Would it also be possible to see one of your example tests? Also, what's a reasonable timeline for a fix? We can do this via chat or a call too if it makes it easier. 😀

**Comment by nwhittaker** [comment]
>…but we might be able to help here by avoiding calls to `setFocus` internally when the host is clicked in some cases.

At minimum it'd be nice to bail early if the element already has focus (if possible). Otherwise we don't really have any way to determine when `setFocus` finishes in that case since we can't wait for an update to `document.activeElement`.

---
>For your use case, is it mainly `calcite-input` wreaking havoc in tests or are there other components that are causing issues?

`calcite-input` was convenient for the test case, but I primarily had problems testing keyboard navigation amongst a collection of `calcite-radio-button` elements (within a custom group element). I suspect it's a potential issue for any component with a `setFocus` implementation, though, due to its waiting for the next animation frame in `componentFocusable`.

---
>Would it also be possible to see one of your example tests?

Most of the test failures revolve around testing keyboard navigation since the browser's logic for determining the next element is dependent on the current `activeElement`.

In this example, the keyboard events were sometimes dispatched to the same element instead of the next one, if I recall correctly. The `waitForFocus` calls were added to work around the regression.

```ts
…
expect(radioA.checked).toBeTruthy()
expect(radioA.tabIndex).toBe(0)

userEvent.keyboard('[ArrowDown]')
await waitForFocus(radioB)
expect(radioA.checked).toBeFalsy()
expect(radioA.tabIndex).toBe(-1)
expect(radioB.checked).toBeTruthy()
expect(radioB.tabIndex).toBe(0)

userEvent.keyboard('[ArrowDown]')
await waitForFocus(radioC)
…

async function waitForFocus(element: Element) {
  return waitFor(() => {
    if (document.activeElement === element) return
    throw new Error()
  })
}
```

---
>Also, what's a reasonable timeline for a fix?

We have some workarounds in place now, so we're no longer blocked -- but I could see the issue being a gotcha with new features going forward. Happy to discuss more on Teams, if needed -- but I imagine a phased approach could be worth considering:
1. Higher priority issue to bail `setFocus` early in cases where focus is delegated(?) or the component already has focus.
2. Lower priority issue to explore eliminating animation frame waiting -- or clarify why/when it's necessary and possibly limiting it to those specific cases.

**Comment by geospatialem** [comment]
A spike effort will be conducted early in 2024 to help determine a timing estimate, with the goal to include prior to the February schedule if in alignment.

**Comment by jcfranco** [comment]
@nwhittaker I'm getting back to this. Given the async nature of components, using `requestAnimationFrame` or a similar delay to ensure focus logic has completed would be the recommended approach.

**WRT proposed approaches:**

> 1. Higher priority issue: bail early in `setFocus` if focus is delegated (possibly?) or the element already has focus.

Even though there's no visual difference on programmatically focusing an already focused element, bailing when focused would break the default scrolling behavior ([example](https://codepen.io/jcfranco/pen/OPJorqr)).

> 2. Lower priority issue: explore eliminating the `requestAnimationFrame` wait, or clarify why/when it's necessary, possibly limiting it to specific cases.

We need focusing to be async to ensure components have completed any rendering updates that could interfere with focus. Even if we bailed on some of these checks (e.g., `componentOnReady`, or resolving immediately when no updates are pending), the parent `setFocus` call would still remain async.

**Additional notes:**

* Focus calls seem to resolve faster now. The test case behaves differently in 3.x and no longer steals focus.
* https://github.com/Esri/calcite-design-system/pull/11825 may improve timing, so the `requestAnimationFrame` approach can work more reliably.

**Comment by nwhittaker** [comment]
>Even though there's no visual difference on programmatically focusing an already focused element, bailing when focused would break the default scrolling behavior ([example](https://codepen.io/jcfranco/pen/OPJorqr)).

I don't think this is an equivalent example. When you click the button, you're taking focus away from the input and then reapplying it. So no need to bail in this scenario.

A more representative example would be to call `focus()` again on the input while it has focus. In this [example][1], you'll see the call to `focusInput()` also scrolls the input back out of view and then calls `focus()` again after 1s which does not cause the element to get scrolled back into view. Since the native input seems to bail early if it already has focus, it seems reasonable that Calcite's elements could behave the same.

>We need focusing to be async to ensure components have completed any rendering updates that could interfere with focus.

I don't think this needs to mean you can't avoid the race condition. Treating `setFocus()` as a cancelable operation should allow for the last call to win.

Specifically, I'm imagining the components' `focus` handler could create a promise or `AbortController` that rejects/aborts if any `focusout` event occurs on it before the component finishes setting focus. Maybe roughly something like:

```ts
class ComponentWithInput {
  abortController: AbortController

  abortFocus: () => {
    this.abortController.abort()
  }

  @Listen('focus')
  onFocus() {
    this.abortController = new AbortController()
    this.el.addEventListener('focusout', this.abortFocus, { once: true })
  }

  async setFocus() {
    await …

    if (!this.abortSignal.aborted) {
      this.el.removeEventListener('focusout', this.abortFocus)
      this.input.focus()
    }
  }
}
```

[1]: https://codepen.io/nwhittaker-esri/pen/vEYVgvG

**Comment by jcfranco** [comment]
> I don't think this is an equivalent example. When you click the button, you're taking focus away from the input and then reapplying it. So no need to bail in this scenario.

Good point. I wasn't taking that into consideration, thanks for the catch.

> I don't think this needs to mean you can't avoid the race condition. Treating setFocus() as a cancelable operation should allow for the last call to win.
> 
> Specifically, I'm imagining the components' focus handler could create a promise or AbortController that rejects/aborts if any focusout event occurs on it before the component finishes setting focus. Maybe roughly something like: 

Ooh, I like this approach and I think it'll fit well for some upcoming controller changes related to `setFocus`. Thanks for sharing. 

Since `setFocus` would still be async, I believe we'd still have the issue where `component.setFocus` and native `focus` behavior will differ in terms of timing (based on the original repro case).

**Comment by jcfranco** [comment]
Quick update: I've [got a PR](https://github.com/Esri/calcite-design-system/pull/11943/files#diff-e57db89670fa4c5dfba71bc71dbf593c2a5b28dd80b84977c2de63f2099bc33a) with a new controller for consistent focus behavior across all components. It's close to landing, it just needs a review and some adjustments to address a few test failures.

Totally optional, but I'd appreciate it if you could take a look at the [spec tests](https://github.com/Esri/calcite-design-system/pull/11943/files#diff-e57db89670fa4c5dfba71bc71dbf593c2a5b28dd80b84977c2de63f2099bc33a) to confirm your use case will be covered. Thanks!

**Comment by nwhittaker** [comment]
Looks great to me, thanks @jcfranco!

**Comment by jcfranco** [comment]
This should be addressed by #11943's refactor.

**Comment by geospatialem** [comment]
Verified with `3.3.0-next.48` and https://codepen.io/geospatialem/pen/vENXLBL using the native `focus` and Calcite's `setFocus` methods.
